### PR TITLE
3744 Add `save_state` utility to handle saving logic

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #3744 .

### Description
Currently, we already have the `copy_model_state` utility to help load `state_dict`, this PR added the `save_state` utility to help simplify the model saving logic. Users easy to face the multi-gpu saving issue (like `state_dict.module` in #3744 ) and may need to save multiple objects in a dict.
And we also received the requests for such a saving utility from NVIDIA Clara users, especially when saving model in BYOW logic without ignite, etc.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/networks/utils.py [hunk 1]
```diff
@@ -436,6 +439,58 @@ def copy_model_state(
...
+    Get the state dict of input object if has `state_dict`, otherwise, return object directly.
+    For data parallel model, automatically convert it to regular model first.
+
+    Args:
+        obj: input object to check and get the state_dict.
+
+    """
+    if isinstance(obj, (nn.DataParallel, nn.parallel.DistributedDataParallel)):
+        obj = obj.module
+    return obj.state_dict() if hasattr(obj, "state_dict") else obj  # type: ignore
```
Comment (wyli) [inline monai/networks/utils.py:453]: thanks! please help update this one: https://github.com/Project-MONAI/MONAI/blob/f08f1d3de838bd3674a3696387447624fcc0cf1f/monai/networks/utils.py#L406-L411

**Review by wyli — APPROVED** [review]

File: monai/networks/utils.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/networks/utils.py]: Sure, updated in the latest commit.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

I totally updated the `saving` logic, could you please help review it again?

Thanks in advance.

File: monai/data/dataset.py [hunk 2]
```diff
@@ -337,27 +344,17 @@ def _cachecheck(self, item_transformed):
         _item_transformed = self._pre_transform(deepcopy(item_transformed))  # keep the original hashed
         if hashfile is None:
             return _item_transformed
-        try:
```
Comment (wyli) [inline monai/data/dataset.py:340]: Not sure if we should unify the saving function here, saving the model state and saving the cache would be very different

File: monai/utils/misc.py [hunk 3]
```diff
@@ -386,3 +391,51 @@ def sample_slices(data: NdarrayOrTensor, dim: int = 1, as_indices: bool = True,
...
+        kwargs: other args for the save `func` except for the checkpoint and filename.
+            default `func` is `torch.save()`, details of other args:
+            https://pytorch.org/docs/stable/generated/torch.save.html.
+
+    """
+    path = Path(path)
+    path_dir = path.parent
+    if not path_dir.exists():
+        if create_dir:
+            path_dir.mkdir(parents=True)
```
Comment (wyli) [inline monai/utils/misc.py:420]: Might have exceptions here when creating

File: monai/utils/misc.py [hunk 4]
```diff
@@ -386,3 +391,51 @@ def sample_slices(data: NdarrayOrTensor, dim: int = 1, as_indices: bool = True,
...
+        else:
+            raise ValueError(f"the directory of specified path is not existing: {path_dir}.")
+    if path.exists():
+        # remove the existing file
+        os.remove(path)
```
Comment (wyli) [inline monai/utils/misc.py:425]: This might raise exceptions

**Review by wyli — CHANGES_REQUESTED** [review]

File: monai/networks/utils.py [hunk 5]
```diff
@@ -436,6 +448,51 @@ def copy_model_state(
...
+    """
+
+    ckpt: Dict = {}
+    if isinstance(src, dict):
+        for k, v in src.items():
+            ckpt[k] = get_state_dict(v)
+    else:
+        ckpt = get_state_dict(src)
+
+    save_obj(obj=ckpt, path=path, create_dir=create_dir, atomic=atomic, func=func, **kwargs)
```
Comment (wyli) [inline monai/networks/utils.py:493]: looks like all the keywords arguments except for `ckpt` could be merged into `kwargs`

File: monai/data/dataset.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/data/dataset.py]: Sure, restored the `PersistentDataset` in the latest commit.

Thanks.

File: monai/networks/utils.py (hunk 5 above)
Comment (Nic-Ma) [reply monai/networks/utils.py:493]: updated in the latest commit.
I kept `obj` and `path` as these 2 args don't have default values, users must provide them.

Thanks.

File: monai/utils/misc.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/utils/misc.py:420]: Hi @wyli ,

May I know which exception when `creating` you mean we should just bypass?
If the regular error, maybe we should let it raise the exception?

Thanks.

**Pull request merged**